internal/authentication/azure/auth.go

```diff
@@ -105,94 +105,158 @@ func (s EnvironmentSettings) GetAzureEnvironment() (*cloud.Configuration, error)
 
-		useTimeout := true
-		if _, ok := os.LookupEnv(identityEndpoint); ok {
-			// App Service & Service Fabric
+	useTimeout := true
+	if _, ok := os.LookupEnv(identityEndpoint); ok {
+		// App Service, Functions, Service Fabric and Container Apps
+		useTimeout = false
+	} else {
+		if _, ok := os.LookupEnv(arcIMDSEndpoint); ok {
+			// Azure Arc
 			useTimeout = false
 		} else {
-			if _, ok := os.LookupEnv(arcIMDSEndpoint); ok {
-				// Azure Arc
+			if _, ok := os.LookupEnv(msiEndpoint); ok {
+				// Cloud Shell
+				useTimeout = false
+			} else if isVirtualMachineWithManagedIdentity() {
+				// Azure VM with MSI enabled
 				useTimeout = false
-			} else {
-				if _, ok := os.LookupEnv(msiEndpoint); ok {
-					// Cloud Shell
-					useTimeout = false
-				} else if isVirtualMachineWithManagedIdentity() {
-					// Azure VM with MSI enabled
-					useTimeout = false
-				}
 			}
 		}
+	}
 
-		// We need to use a timeout for MSI on environments where it is not available because the request for the default IMDS endpoint can hang for several minutes.
-		if useTimeout {
-			msiCred = &timeoutWrapper{cred: msiCred, authmethod: "managed identity", timeout: 1 * time.Second}
-		}
+	// We need to use a timeout for MSI on environments where it is not available because the request for the default IMDS endpoint can hang for several minutes.
+	if useTimeout {
+		msiCred = &timeoutWrapper{cred: msiCred, authmethod: "managed identity", timeout: timeout}
+	}
 
-		if err == nil {
-			creds = append(creds, msiCred)
-		} else {
-			errs = append(errs, err)
-		}
+	if err == nil {
+		*creds = append(*creds, msiCred)
+	} else {
+		*errs = append(*errs, err)
 	}
+}
 
-	// 5. AzureCLICredential
-	{
-		cred, credErr := azidentity.NewAzureCLICredential(nil)
-		if credErr == nil {
-			creds = append(creds, &timeoutWrapper{cred: cred, authmethod: "Azure CLI", timeout: 30 * time.Second})
-		} else {
-			errs = append(errs, credErr)
+func (s EnvironmentSettings) addCLIProvider(timeout time.Duration, creds *[]azcore.TokenCredential, errs *[]error) {
+	cred, credErr := azidentity.NewAzureCLICredential(nil)
+	if credErr == nil {
+		*creds = append(*creds, &timeoutWrapper{cred: cred, authmethod: "Azure CLI", timeout: 30 * time.Second})
+	} else {
+		*errs = append(*errs, credErr)
+	}
+}
+
+func (s EnvironmentSettings) addProviderByAuthMethodName(authMethod string, creds *[]azcore.TokenCredential, errs *[]error) {
+	switch authMethod {
+	case "serviceprincipal":
+		s.addClientCredentialsProvider(creds, errs)
+	case "certificate":
+		s.addClientCertificateProvider(creds, errs)
+	case "workloadidentity":
+		s.addWorkloadIdentityProvider(creds, errs)
+	case "managedidentity":
+		s.addManagedIdentityProvider(1*time.Second, creds, errs)
+	case "cli":
+		s.addCLIProvider(30*time.Second, creds, errs)
+	}
+}
+
+func getAzureAuthMethods() [6]string {
+	return [...]string{"serviceprincipal", "certificate", "workloadidentity", "managedidentity", "cli", "none"}
```

Comment:
Just return a slice?
```suggestion
func getAzureAuthMethods() []string {
	return []string{"serviceprincipal", "certificate", "workloadidentity", "managedidentity", "cli", "none"}
```

Reply:
It's silly optimization but this makes the compiler allocate a fixed length slice. So why not. I'll keep it, don't want anyone to dynamically append to this.